src/mechanics/mesh/MeshGmsh.cpp

```diff
@@ -154,49 +176,150 @@ std::vector<GmshElement> ReadElementsASCII(std::ifstream& rFile)
+         * if you are interested.
+         *
+         * TLDR: The element section looks as follows:
+         * number-of-elements
+         * element-header-binary
+         * elements-binary
+         * element-header-binary
+         * elements-binary
+         * ...
+         *
+         * element-header-binary contains 3 ints
+         *  [element_type, num_elm_follow, num_tags]
+         *
+         * So we have an extra loop (indexH) over all elements with the same header
+         */
+
+        rFile.read((char*)&element_type, sizeof(int));
+        rFile.read((char*)&num_elm_follow, sizeof(int)); // number of elements with the same header
+        rFile.read((char*)&num_tags, sizeof(int));
+
+        int cur_num_elm_nodes = NumNodes(element_type);
+
+        for (int indexH = 0; indexH < num_elm_follow; indexH++)
+        {
+
+            // set element type
+            elements[elemCount].type = element_type;
+
+            // read element number
+            rFile.read((char*)&elements[elemCount].id, sizeof(int));
+
+            elements[elemCount].tags.resize(num_tags);
+            elements[elemCount].nodes.resize(cur_num_elm_nodes);
+
+            // read tags
+            for (int& tag : elements[elemCount].tags)
+                rFile.read((char*)&tag, sizeof(int));
+
+            // read nodes
+            for (int& node : elements[elemCount].nodes)
+                rFile.read((char*)&node, sizeof(int));
+
+            elemCount += indexH;
+        }
+    }
+    getline(rFile, line); // endl
+    ExpectNextLineToBe(rFile, "$EndElements");
+    return elements;
+}
+
+
+std::vector<GmshPhysicalNames> ReadPhysicalNames(std::ifstream& rFile)
 {
     std::string line;
     std::getline(rFile, line);
     int numNames = std::stoi(line);
 
     std::vector<GmshPhysicalNames> physicalNames(numNames);
-    for (GmshPhysicalNames& name : physicalNames)
+    for (GmshPhysicalNames& physicalName : physicalNames)
     {
-        rFile >> name.dimension;
-        rFile >> name.id;
-        rFile >> name.name;
+        rFile >> physicalName.dimension;
+        rFile >> physicalName.id;
+        rFile >> physicalName.physicalName;
 
         // remove quotation marks from string
-        name.name.erase(std::remove(name.name.begin(), name.name.end(), '\"'), name.name.end());
+        physicalName.physicalName.erase(
+                std::remove(physicalName.physicalName.begin(), physicalName.physicalName.end(), '\"'),
+                physicalName.physicalName.end());
         std::getline(rFile, line);
     }
     ExpectNextLineToBe(rFile, "$EndPhysicalNames");
     return physicalNames;
 }
 
-void ProcessSectionASCII(std::ifstream& rFile, GmshFileContent& rFileContent)
+void ProcessSection(std::ifstream& rFile, GmshFileContent& rFileContent)
 {
     std::string line;
     std::getline(rFile, line);
-    if (line[0] == '$')
+
+    if (line.empty())
+        return;
+
+    if (line == "$PhysicalNames")
     {
```

Comment:
By just checking equality (you removed the compare) trailing whitespaces in the Gmsh file will produce errors. Gmsh itself has no problem with that so NuTo also shouldn't (IMHO). This is an error that is I think not easy to find, so why not just use compare?

Reply:
`string1.compare(string2)` is actually equal to `string1 == string2`. But I solved that trailing whitespace issue by comparing both strings only up to `string2.size()`.